File: nodejs/mailer/consumer.js
const kafkaClient = require("promise-node-rdkafka");
const bunyan = require("bunyan");

const mailSender = require("./mail-sender");

const logger = bunyan.createLogger({ name: "mailer-consumer" });

const consumer = new kafkaClient.PromiseConsumer({
    "bootstrap.servers": "kafka:9092",
    "group.id": "mailer",
    "enable.auto.commit": false    
}, {
    "auto.offset.reset": "earliest"
});

let running = true;

(async () => {
    await consumer.connect();
    consumer.subscribe(["send-mail"]);
    try {
        while(running) {
            const messages = await consumer.consume(1);
            if (messages.length > 0) {
                const to = messages[0].key.toString();
                const msg = JSON.parse(messages[0].value.toString());
                
                logger.info("Sending mail:", to, msg);
                
                await mailSender(to, msg["subject"], msg["text"]);
                consumer.commit();
            }
        }
    } finally {
        await consumer.disconnect(10000);
    }

})().catch((err) => {
    logger.error(err);
});

process.on("SIGINT", () => { running = false; })

File: python/basic_kafka/simple/consumer.py
from confluent_kafka import Consumer

def main():
    consumer = Consumer({
        'bootstrap.servers': 'kafka:9092',
        'group.id': 'logger',
        'enable.auto.commit': False,
        'default.topic.config': {'auto.offset.reset': 'smallest'}
    })
    consumer.subscribe(['test'])

    try:
        while True:
            msg = consumer.poll(timeout=1.0)
            if msg:
                print(msg.key() if msg.key() else 'No-key', msg.value().decode())
                consumer.commit()
    finally:
        consumer.close()


if __name__ == '__main__':
    main()


